app/backend/src/couchers/servicers/api.py

```diff
@@ -429,23 +429,44 @@ def ListFriends(self, request, context):
             )
 
     def SendFriendRequest(self, request, context):
-        with session_scope() as session:
-            from_user = session.query(User).filter(User.id == context.user_id).one_or_none()
-
-            if not from_user:
-                context.abort(grpc.StatusCode.NOT_FOUND, errors.USER_NOT_FOUND)
+        if context.user_id == request.user_id:
+            context.abort(grpc.StatusCode.FAILED_PRECONDITION, errors.CANT_FRIEND_SELF)
 
+        with session_scope() as session:
+            user = session.query(User).filter(User.id == context.user_id).one()
             to_user = session.query(User).filter(User.id == request.user_id).one_or_none()
 
             if not to_user:
                 context.abort(grpc.StatusCode.NOT_FOUND, errors.USER_NOT_FOUND)
 
-            if get_friends_status(session, from_user.id, to_user.id) != api_pb2.User.FriendshipStatus.NOT_FRIENDS:
+            if (
+                session.query(FriendRelationship)
+                .filter(
+                    or_(
+                        and_(
+                            FriendRelationship.from_user_id == context.user_id,
+                            FriendRelationship.to_user_id == request.user_id,
+                        ),
+                        and_(
+                            FriendRelationship.from_user_id == request.user_id,
+                            FriendRelationship.to_user_id == context.user_id,
+                        ),
+                    )
+                )
+                .filter(
+                    or_(
+                        FriendRelationship.status == FriendStatus.accepted,
+                        FriendRelationship.status == FriendStatus.pending,
+                    )
+                )
+                .one_or_none()
+                is not None
+            ):
                 context.abort(grpc.StatusCode.FAILED_PRECONDITION, errors.FRIENDS_ALREADY_OR_PENDING)
 
-            # Race condition!
+            # TODO: Race condition where we can create two friend reqs, needs db constraint! See comment in table
```

Comment:
Aside from the race condition, what happens in the case where the request exists but was cancelled/rejected?

Reply:
It shouldn't change the backend behaviour

Reply:
Aren't we making a duplicate friend relationship?

Reply:
Yep